packages/veritone-redux-common/src/helpers/redux/promiseMiddleware.js

```diff
@@ -0,0 +1,114 @@
+export const ERROR_ACTION = Symbol('ERROR_ACTION');
+export const CALLBACK_ARGUMENT = Symbol('CALLBACK_ARGUMENT');
+export const CALLBACK_ERROR_ARGUMENT = Symbol('ERROR_CALLBACK_ARGUMENT');
+
+export function promiseMiddleware() {
+  let pendingActionList = [];
+  const promisesList = [];
+
+  const getPromisesList = () => {
+    return promisesList;
+  };
+
+  const middleware = store => {
+    return function(next) {
+      return function(action) {
+        const successAction = action[WAIT_FOR_ACTION];
+        const errorAction = action[ERROR_ACTION];
+        const newPendingActionInfo = {};
+
+        // loop through list, see if there are any pending
+        // promise callbacks by comparing types
+        forEachRight(pendingActionList, pendingActionInfo => {
+          if (pendingActionInfo.isSuccessAction(action)) {
+            pendingActionInfo.resolveCallback(
+              pendingActionInfo.successArgumentCb(action)
+            );
+            pendingActionList.splice(
+              pendingActionList.indexOf(pendingActionInfo),
+              1
+            );
+          } else if (pendingActionInfo.isErrorAction(action)) {
+            pendingActionInfo.rejectCallback(
+              pendingActionInfo.errorArgumentCb(action)
+            );
+            pendingActionList.splice(
+              pendingActionList.indexOf(pendingActionInfo),
+              1
+            );
+          }
+        });
+
+        // If new action doesn't contain the symbol then we
+        // are done
+        if (!action[WAIT_FOR_ACTION]) {
+          return next(action);
+        }
+
+        // These are the comparison function
+        newPendingActionInfo.isSuccessAction = isFunction(successAction)
+          ? successAction
+          : action => action.type === successAction;
+        if (errorAction) {
+          newPendingActionInfo.isErrorAction = isFunction(errorAction)
+            ? errorAction
+            : action => action.type === errorAction;
+        } else {
+          newPendingActionInfo.isErrorAction = constant(false);
+        }
+
+        // If you need to change which action info to push into the
+        // promise callbacks, you can change them here
+        newPendingActionInfo.successArgumentCb =
+          action[CALLBACK_ARGUMENT] || fsaCompliantArgumentCb;
+        newPendingActionInfo.errorArgumentCb =
+          action[CALLBACK_ERROR_ARGUMENT] || fsaCompliantErrorArgumentCb;
+
+        // Setting up promise to be called when action types match
+        const promise = new Promise(function(resolve, reject) {
+          newPendingActionInfo.resolveCallback = resolve;
+          newPendingActionInfo.rejectCallback = reject;
+        });
+
+        pendingActionList.push(newPendingActionInfo);
+        promisesList.push(promise);
+
+        // Move on, but return a promise for dispatch
+        next(action);
+        return promise;
+      };
+    };
+  };
+  return Object.assign(middleware, { getPromisesList });
```

Comment:
promisesList/getPromisesList seems to be useless?

Reply:
unless you want the list of promises...  yea it's useless